File: Utils/PadesUtils.m
//
//  PadesUtils.m
//  AutofirmaJ2OBJC
//
//  Created by Desarrollo Abamobile on 14/8/24.
//

#import <Foundation/Foundation.h>
#import "PadesUtils.h"
#import "AOPadesUtils.h"

@implementation PadesUtils

- (void)signPdf:(NSData *)pdfData
  signAlgorithm:(nullable NSString *)signAlgorithm
	privateKey:(SecKeyRef)privateKey
    certificate:(SecCertificateRef)certificateRef
certificateAlgorithm:(nullable NSString *)certificateAlgorithm
    extraParams:(nullable NSDictionary<NSString *, NSString *> *)extraParams
	completion:(SignPdfCompletionHandler)completion {
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
	   AOPadesUtils *utils = [[AOPadesUtils alloc] init];
	   
	   [utils signPdfWithData:pdfData
			  signAlgorithm:signAlgorithm
				privateKey:privateKey
			    certificate:certificateRef
		certificateAlgorithm:certificateAlgorithm
			    extraParams:extraParams
				completion:^(NSString * _Nullable result, NSError * _Nullable error) {
		  dispatch_async(dispatch_get_main_queue(), ^{
			 if (error) {
				completion(nil, error);
			 } else if (result) {
				completion(result, nil);
			 } else {
				NSError *unknownError = [NSError errorWithDomain:@"Error"
												    code:9999
												userInfo:@{NSLocalizedDescriptionKey: @"Unknown error occurred"}];
				completion(nil, unknownError);
			 }
		  });
	   }];
    });
}

@end

File: Utils/PadesUtils.swift
//
//  PadesUtils.swift
//  Autofirma
//
//  Created by Desarrollo Abamobile on 12/8/24.
//  Copyright © 2024 Solid GEAR. All rights reserved.
//

import Foundation
import Security



class PadesUtils {
    func signPdf(
	   pdfData: Data,
	   algorithm: String?,
	   privateKey: SecKey,
	   certificateRef: SecCertificate,
	   extraParams: [String: String]?,
	   completion: @escaping (Result<String, Error>) -> Void
    ) {
	   DispatchQueue.global(qos: .userInitiated).async {
		  let utils = AOPadesUtils()

		  utils.signPdf(
			 with: pdfData,
			 algorithm: algorithm,
			 privateKey: privateKey,
			 certificate: certificateRef,
			 extraParams: extraParams
		  ) { result, error in
			 DispatchQueue.main.async {
				if let error = error {
				    completion(.failure(error))
				} else if let result = result {
				    completion(.success(result))
				} else {
				    let unknownError = NSError(domain: "Error", code: 9999, userInfo: [NSLocalizedDescriptionKey: "Unknown error occurred"])
				    completion(.failure(unknownError))
				}
			 }
		  }
	   }
    }
}
